xbmc/addons/VFSEntry.cpp

```diff
@@ -0,0 +1,666 @@
+  CVFSEntryIFileWrapper(ptr)
+{
+}
+
+bool CVFSEntryILiveTVWrapper::NextChannel(bool preview)
+{
+  if (!m_context)
+    return false;
+
+  return m_addon->NextChannel(m_context, preview);
+}
+
+bool CVFSEntryILiveTVWrapper::PrevChannel(bool preview)
+{
+  if (!m_context)
+    return false;
+
+  return m_addon->PrevChannel(m_context, preview);
+}
+
+bool CVFSEntryILiveTVWrapper::SelectChannel(unsigned int channel)
+{
+  if (!m_context)
+    return false;
+
+  return m_addon->SelectChannel(m_context, channel);
+}
+
+int CVFSEntryILiveTVWrapper::GetTotalTime()
+{
+  if (!m_context)
+    return 0;
+
+  return m_addon->GetTotalTime(m_context);
+}
+
+int CVFSEntryILiveTVWrapper::GetStartTime()
+{
+  if (!m_context)
+    return 0;
+
+  return m_addon->GetStartTime(m_context);
+}
+
+bool CVFSEntryILiveTVWrapper::UpdateItem(CFileItem& item)
+{
+  if (!m_context)
+    return false;
+
+  return m_addon->UpdateItem(m_context, item);
+}
+
+CVFSEntryIDirectoryWrapper::CVFSEntryIDirectoryWrapper(VFSEntryPtr ptr) :
+  m_addon(ptr)
+{
+}
+
+bool CVFSEntryIDirectoryWrapper::Exists(const CURL& url)
+{
+  return m_addon->DirectoryExists(url);
+}
+
+bool CVFSEntryIDirectoryWrapper::Remove(const CURL& url)
+{
+  return m_addon->RemoveDirectory(url);
+}
+
+bool CVFSEntryIDirectoryWrapper::Create(const CURL& url)
+{
+  return m_addon->CreateDirectory(url);
+}
+
+bool CVFSEntryIDirectoryWrapper::GetDirectory(const CURL& url,
+                                              CFileItemList& items)
+{
+  return m_addon->GetDirectory(url, items, this);
+}
+
+bool CVFSEntryIDirectoryWrapper::DoGetKeyboardInput(void* ctx,
+                                                    const char* heading,
+                                                    char** input)
+{
+  return ((CVFSEntryIDirectoryWrapper*)ctx)->GetKeyboardInput2(heading, input);
+}
+
+bool CVFSEntryIDirectoryWrapper::GetKeyboardInput2(const char* heading,
+                                                   char** input)
+{
+  std::string inp;
+  bool result;
+  if ((result=GetKeyboardInput(CVariant(std::string(heading)), inp)))
+    *input = strdup(inp.c_str());
+
+  return result;
+}
+
+void CVFSEntryIDirectoryWrapper::DoSetErrorDialog(void* ctx, const char* heading,
+                                                  const char* line1,
+                                                  const char* line2,
+                                                  const char* line3)
+{
+  ((CVFSEntryIDirectoryWrapper*)ctx)->SetErrorDialog2(heading, line1,
+                                                      line2, line3);
+}
+
+void CVFSEntryIDirectoryWrapper::SetErrorDialog2(const char* heading,
+                                                 const char* line1,
+                                                 const char* line2,
+                                                 const char* line3)
+{
+  CVariant l2=0, l3=0;
+  if (line2)
+    l2 = std::string(line2);
```

Comment:
is line2 null-checked before this line? I traced the execution back to the DLL, so we should sanity-check its values somewhere (probably as soon as it leaves the DLL boundary)